sam/io.go

```diff
@@ -40,17 +40,158 @@ func ReadToChan(filename string, data chan<- Sam, header chan<- Header) {
+		if err == io.EOF {
+			break
+		}
+		sendRecords <- curr
+	}
+
+	err = file.Close()
+	exception.PanicOnErr(err)
+	close(sendRecords)
+}
+
 // GoReadToChan streams the input file so that only a small portion
-// of the file is kept in memory at a time. This function wraps the
-// ReadToChan function to automatically handle channel creation,
-// goroutine spawning, and header retrieval.
+// of the file is kept in memory at a time. This function automatically
+// handles channel creation, goroutine spawning, and header retrieval.
+//
+// GoReadToChan will detect if the input file ends in ".bam" and
+// automatically switch to a bam parser.
 func GoReadToChan(filename string) (<-chan Sam, Header) {
 	data := make(chan Sam, 1000)
 	header := make(chan Header)
-	go ReadToChan(filename, data, header)
+
+	if strings.HasSuffix(filename, ".bam") {
+		go readBamToChan(filename, data, header)
+	} else {
+		go readSamToChan(filename, data, header)
+	}
 	return data, <-header
 }
 
+// GoReadToChanRecycle streams the input file so that only a small portion
+// of the file is kept in memory at a time. This function automatically
+// handles channel creation, goroutine spawning, and header retrieval.
+//
+// GoReadToChanRecycle will detect if the input file ends in ".bam" and
+// automatically switch to a bam parser.
+//
+// Unlike GoReadToChan, GoReadToChanRecycle has 2 channel returns, a
+// receiver channel (parsedRecords) and a sender channel (recycledStructs).
+// Parsed sam records can be accessed from the receiver channel. Once the
+// received struct is no longer needed, it can be returned to this function
+// via the sender channel. Compared to GoReadToChan, this system significantly
+// reduces memory allocations as structs can be reused rather then discarded.
+// The total number of Sam structs that are allocated can be set with the
+// bufferSize input variable.
+//
+// Note that if bufferSize number of structs are retained by the calling function
+// and not returned to this function. No new records will be sent and the
+// goroutines will deadlock.
+//
+// One notable disadvantage is that each read is only valid until the struct
+// is returned to this function. This means that any information that must be
+// retained between reads must be copied by the calling function.
+func GoReadToChanRecycle(filename string, bufferSize int) (parsedRecords <-chan *Sam, recycledStructs chan<- *Sam, header Header) {
+	parsedRecordsInit := make(chan *Sam, bufferSize)
+	recycledStructsInit := make(chan *Sam, bufferSize)
+	headerChan := make(chan Header)
+
+	if strings.HasSuffix(filename, ".bam") {
+		go readBamToChanRecycle(filename, parsedRecordsInit, recycledStructsInit, headerChan)
+	} else {
+		go readSamToChanRecycle(filename, parsedRecordsInit, recycledStructsInit, headerChan)
+	}
+
+	// spawn a temporary goroutine that will allocate bufferSize number of
+	// Sam structs and send them to the readToChan function for reading.
+	go func(chan<- *Sam) {
```

Comment:
Is it easier/simpler to have this not be a go routine and just do a for loop that fills up the buffer?